cpu/native/ng_tapnet/ng_tapnet.c

```diff
@@ -0,0 +1,615 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @ingroup     cpu_native_ng_tapnet
+ * @{
+ *
+ * @file
+ *
+ * @author      Ludwig Ortmann <ludwig.ortmann@fu-berlin.de>
+ * @author      Martine Lenders <mlenders@inf.fu-berlin.de>
+ */
+
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <net/ethernet.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#ifdef __MACH__
+#define _POSIX_C_SOURCE
+#include <net/if.h>
+#undef _POSIX_C_SOURCE
+#include <ifaddrs.h>
+#include <net/if_dl.h>
+
+#elif defined(__FreeBSD__)
+#include <sys/socket.h>
+#include <net/if.h>
+#include <ifaddrs.h>
+#include <net/if_dl.h>
+
+#else
+#include <net/if.h>
+#include <linux/if_tun.h>
+#include <linux/if_ether.h>
+#endif
+
+#include "native_internal.h"
+#include "net/ng_netdev.h"
+#include "net/ng_netif/hdr.h"
+#include "net/ng_pkt.h"
+#include "net/ng_pktbuf.h"
+#include "thread.h"
+#include "utlist.h"
+
+#include "ng_tapnet.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _ISR_EVENT_RX   (1U)
+
+extern ng_tapnet_t ng_tapnet;
+
+#ifdef __MACH__
+pid_t _sigio_child_pid;
+
+static void _sigio_child(void)
+{
+    pid_t parent = _native_pid;
+
+    if ((_sigio_child_pid = real_fork()) == -1) {
+        err(EXIT_FAILURE, "sigio_child: fork");
+    }
+
+    if (_sigio_child_pid > 0) {
+        /* return in parent process */
+        return;
+    }
+
+    /* watch tap interface and signal parent process if data is
+     * available */
+
+    fd_set rfds;
+
+    while (1) {
+        FD_ZERO(&rfds);
+        FD_SET(_native_tap_fd, &rfds);
+
+        if (real_select(_native_tap_fd + 1, &rfds, NULL, NULL, NULL) == 1) {
+            kill(parent, SIGIO);
+        }
+        else {
+            kill(parent, SIGKILL);
+            err(EXIT_FAILURE, "osx_sigio_child: select");
+        }
+
+        pause();
+    }
+}
+#endif
+
+static int _marshall_ethernet(ng_tapnet_t *dev, uint8_t *buffer, ng_pktsnip_t *pkt)
```

Comment:
Reversed packet order for sending in `ng_pktbuf` makes this very clunky, since I basically need to start at the end of a single linked list and than have to work my way up to the front. Any idea @haukepetersen?